### PR TITLE
Integrating mail notifications with push notifications

### PR DESCRIPTION
**Pull request by notbakaneko**
- Switches sending of mail notifications to using `user_notifications` from the push notification system.
- notification mail now batched and sent periodically as a digest.
- more delivery options added since additional notifications get mailed now.
- some of the digest grouping probably needs working on 🤔 
- it doesn't do any last visited checking yet.

`NotificationsSendMail` command collects a list of users that need mail notification digests sent and then queues `Jobs\UserNotificationDigest` jobs up per user. `Jobs\UserNotificationDigest` gets the actual list of notifications to send and runs the mailer, or aborts if there aren't actually any notifications to send.
`Mail\UserNotificationDigest` is intentionally not queueable and `Jobs\UserNotificationDigest` has a minimal number of fields and does not store notifications to avoid repeatedly serializing and unserializing models with the queue.

This isn't the same as having detailed mail notifications on the level of push notifications as there are other issues we'd want to consider for going in that direction (having another entire set of renderers just for detailed mail notifications is one..).

Old mailer classes kept temporarily for now so already queued jobs don't explode on deploy.
- [x] should be setting last sent id on migration
- [x] `2020_06_12_112517_add_delivery_to_user_notifications`

### CODE REVIEW

File: app/Jobs/Notifications/ChannelMessage.php [hunk 1]
```diff
@@ -6,12 +6,26 @@
...
 
 class ChannelMessage extends BroadcastNotificationBase
 {
+    const NOTIFICATION_OPTION_NAME = Notification::CHANNEL_MESSAGE;
+
     protected $message;
 
+    public static function getBaseKey(Notification $notification): string
+    {
+        return 'channel.channel.pm';
```
Comment (nanaya) [inline app/Jobs/Notifications/ChannelMessage.php:20]: but is it always pm

File: app/Jobs/UserNotificationDigest.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        $this->user = $user;
+        $this->fromId = $fromId;
+        $this->toId = $toId;
+    }
+
+    public function handle()
+    {
+        $notifications = Notification
+            ::whereHas('userNotifications', function ($q) {
+                $q->where('user_id', $this->user->getKey())->hasMailDelivery();
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Jobs/UserNotificationDigest.php:35]: I'm not sure if notifications going out through a different delivery method should not be sending read notifications 🤔 
If someone opts to receive notifications be several methods, usually they'd expect to get all of them?

[the remark this answers is not in the document]
Comment (nanaya) [reply app/Jobs/UserNotificationDigest.php:35]: Maybe if it's individual, instant notification. As for digest, sending read notification which possibly up to 10 minutes late, doesn't seem that useful.

**Comment by nanaya** [comment]
Should notification bundle queries be updated to exclude non-push notifications?

File: resources/lang/en/notifications.php [hunk 3]
```diff
@@ -133,4 +133,40 @@
...
+            'beatmapset_discussion_lock' => 'The discussion on ":title" has been locked',
+            'beatmapset_discussion_unlock' => 'The discussion on ":title" has been unlocked',
+            'beatmapset_problem' => 'A new problem reported was reported on ":title"',
+            'beatmapset_state' => 'The state of ":title" has changed:',
+            'comment' => '":title" has new comments:',
+        ],
+
+        'channel' => [
+            'channel' => [
+                'pm' => 'You\'ve received a new message from :username',
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/lang/en/notifications.php:149]: but there'll always be at least one 🤔

File: app/Jobs/Notifications/BroadcastNotificationBase.php [hunk 4]
```diff
@@ -92,26 +125,29 @@ public function getTimestamp()
...
+        // so, it should be fine to create the user notifications first.
 
-        DB::transaction(function () use ($notification, $receiverIds) {
-            foreach ($receiverIds as $id) {
-                $notification->userNotifications()->create(['user_id' => $id]);
+        $pushReceiverIds = [];
+        $notification->getConnection()->transaction(function () use ($deliverySettings, $notification, &$pushReceiverIds) {
+            foreach ($deliverySettings as $userId => $delivery) {
+                $userNotification = $notification->userNotifications()->create(['delivery' => $delivery, 'user_id' => $userId]);
+                $userNotification->isPush() && $pushReceiverIds[] = $userId;
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Jobs/Notifications/BroadcastNotificationBase.php:144]: Now I'm wondering if the notification history page should be showing all notifications or just push notifications 🤔

**Comment by nanaya** [comment]
conflicts

**Comment by peppy** [comment]
That migration needs online execution, correct?

**Comment by nanaya** [comment]
Yeah, namely column `delivery` of unsigned tinyint default 1 after column `is_read` on table `user_notifications`.

```sql
ALTER TABLE user_notifications ADD delivery tinyint unsigned NOT NULL DEFAULT 1 AFTER delivery;
```

Then the related entry `2020_06_12_112517_add_delivery_to_user_notifications` in `migrations` table.

**Comment by peppy** [comment]
Database migrations done.

**Pull request merged**